# 01_04 Pandas

Pandas는 numpy가 해결해 주지 못하는 고수준의 API를 사용하게 해주는 패키지입니다.<br>
대부분의 데이터는 2차원 (행, 열)로 이루어져 있으며, 따라서 DB의 Table이나 Excel의 sheet로 데이터를 관리하게 됩니다.<br>
Pandas는 이러한 2차원 데이터를 효율적으로 가공/처리하는 기능을 제공합니다.

RDBMS, Excel 등 보다는 덜 할 수 있어도, Pandas는 Python 안에서는 이에 비준하는 기능을 제공합니다.

## DataFrame

DataFrame은 Pandas의 핵심 객체로서, 2차원 데이터를 담는 데이터 구조체입니다.<br>
DataFrame은 Index, Series로 이루어져 있습니다.

- Index: RDBMS의 고유 ID, Key값을 의미합니다.
- Series: Column이 하나인 데이터 구조체입니다.

따라서 DataFrame은 여러 Series로 구성된 하나의 데이터 구조체(테이블)이 됩니다.

Pandas는 pd로 보통 pd로 줄여서 많이 사용합니다.

In [1]:
import pandas as pd

먼저 Pandas로 csv파일을 읽어오도록 하겠습니다.<br>
데이터 셋은 Kaggle에서 불러올 예정이라 다음 링크에서 데이터를 다운받아주시면 됩니다.<br>
https://www.kaggle.com/c/titanic/data

Data Explorer에서 [Download All] 버튼을 눌러 원하는 경로에 저장합니다.<br>
저는 모든 데이터를 root 경로의 `data` 폴더에 저장할 예정이므로 경로가 다를 수 있습니다.

In [7]:
!ls ../data
!ls ../data/titanic

titanic
gender_submission.csv test.csv              train.csv


In [8]:
import os

TITANIC_DIR = "../data/titanic"
TITANIC_TRAIN_PATH = os.path.join(TITANIC_DIR, "train.csv")
TITANIC_TEST_PATH = os.path.join(TITANIC_DIR, "test.csv")

### 데이터 파일 읽어오기

Pandas는 csv, table, fwf 등 다양한 파일을 읽어올 수 있습니다.
- `read_csv()`
- `read_table()`
- `read_fwf()`

이러한 파일이 아니더라도, 규칙성이 있는 텍스트 기반 파일이라면 Delimeter를 지정하여 DataFrame으로 변환이 가능합니다.<br>
`read_csv()` 함수의 `seq=`에 해당 Delimeter를 지정하여 가져옵니다.

In [9]:
titanic_df = pd.read_csv(TITANIC_TRAIN_PATH)
# head는 index 기준 최상위 5개를 가져옵니다.
titanic_df.head(5)

,PassengerId,Survived,Pclass,Name,Sex,Age,SibSp,Parch,Ticket,Fare,Cabin,Embarked
0,1,0,3,"Braund, Mr. Owen Harris",male,22.0,1,0,A/5 21171,7.2500,NaN,S
1,2,1,1,"Cumings, Mrs. John Bradley (Florence Briggs Th...",female,38.0,1,0,PC 17599,71.2833,C85,C
2,3,1,3,"Heikkinen, Miss. Laina",female,26.0,0,0,STON/O2. 3101282,7.9250,NaN,S
3,4,1,1,"Futrelle, Mrs. Jacques Heath (Lily May Peel)",female,35.0,1,0,113803,53.1000,C123,S
4,5,0,3,"Allen, Mr. William Henry",male,35.0,0,0,373450,8.0500,NaN,S


In [10]:
print(f"titanic_df type: {type(titanic_df)}")

titanic_df type: <class 'pandas.core.frame.DataFrame'>


DataFrame의 shape 또한 `shape`로 불러올 수 있습니다.

In [11]:
print(f"titanic_df shape: {titanic_df.shape}")

titanic_df shape: (891, 12)


이 데이터 셋의 구성은 12개짜리 Feature를 가진 891개의 데이터가 됩니다.

DataFrame은 다음 정보도 확인 할 수 있습니다.
- Column type
- Null 데이터 개수
- 데이터 분포도 등..

#### DataFrame의 정보 조회 - info()

In [12]:
titanic_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 891 entries, 0 to 890
Data columns (total 12 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   PassengerId  891 non-null    int64  
 1   Survived     891 non-null    int64  
 2   Pclass       891 non-null    int64  
 3   Name         891 non-null    object 
 4   Sex          891 non-null    object 
 5   Age          714 non-null    float64
 6   SibSp        891 non-null    int64  
 7   Parch        891 non-null    int64  
 8   Ticket       891 non-null    object 
 9   Fare         891 non-null    float64
 10  Cabin        204 non-null    object 
 11  Embarked     889 non-null    object 
dtypes: float64(2), int64(5), object(5)
memory usage: 83.7+ KB


각 컬럼의 구성과 결측치, data type 등을 한 번에 출력합니다.<br>
전체 891개 데이터 중 결측치가 Age(891-714=177개), Cabin(891-204=687개)에 있다는 것을 확인하였습니다.

#### DataFrame의 수치형 데이터 분포도 - describe()

`describe()`는 계산이 가능한 경우에만 표시되므로 `dtype`이 숫자형이 아니라면 표시되지 않습니다.

In [13]:
titanic_df.describe()

,PassengerId,Survived,Pclass,Age,SibSp,Parch,Fare
count,891.000000,891.000000,891.000000,714.000000,891.000000,891.000000,891.000000
mean,446.000000,0.383838,2.308642,29.699118,0.523008,0.381594,32.204208
std,257.353842,0.486592,0.836071,14.526497,1.102743,0.806057,49.693429
min,1.000000,0.000000,1.000000,0.420000,0.000000,0.000000,0.000000
25%,223.500000,0.000000,2.000000,20.125000,0.000000,0.000000,7.910400
50%,446.000000,0.000000,3.000000,28.000000,0.000000,0.000000,14.454200
75%,668.500000,1.000000,3.000000,38.000000,1.000000,0.000000,31.000000
max,891.000000,1.000000,3.000000,80.000000,8.000000,6.000000,512.329200


- count: 데이터의 갯수
- mean: 총 데이터의 평균
- std: 총 데이터의 표준편차
- min: 총 데이터 중 최솟값
- 25%: 총 데이터 중 25%에 해당하는 값
- 50%: 총 데이터 중 50%에 해당하는 값
- 75%: 총 데이터 중 75%에 해당하는 값
- max: 총 데이터 중 최댓값

해당 출력에서 확인할 수 있는 사항은 다음과 같습니다.
- PassengerID는 ID 값이므로 큰 의미가 없습니다.
- Survived의 min이 0, max가 1이므로 0(Dead), 1(Survived)로 나뉘어진 범주형 데이터입니다.
- Pclass의 min이 1, max가 3이므로 1, 2, 3등급으로 나뉘어진 범주형 데이터입니다.
- Age를 알 수 없는 승객 데이터가 존재합니다.

#### DataFrame의 범주형 데이터 분포도 - value_counts()

`describe()`가 수치형 데이터의 분포도를 표시해준다면, `value_counts()`는 범주형 데이터의 분포도를 확인하는데 사용합니다.<br>
해당 함수는 Series를 반환합니다.<br>
다만 컬럼을 지정하여 사용해야하기 때문에, 위에서 분석한 Survived, Pclass를 예시로 불러오도록 하겠습니다.

In [16]:
survived_counts = titanic_df["Survived"].value_counts()
print(f"value_counts type: {type(survived_counts)}")
print(f"생존 인원 통계:\n{survived_counts}")

value_counts type: <class 'pandas.core.series.Series'>
생존 인원 통계:
Survived
0    549
1    342
Name: count, dtype: int64


In [19]:
pclass_counts = titanic_df["Pclass"].value_counts()
print(f"승객 등급 통계:\n{pclass_counts}")

승객 등급 통계:
Pclass
3    491
1    216
2    184
Name: count, dtype: int64


#### 결측치 무시 - dropna()

`value_counts()`는 결측치를 포함한 결과값을 출력합니다.<br>
예시로 Embarked column을 확인해보겠습니다.

In [32]:
print(f"타이타닉 데이터 갯수: {titanic_df.shape[0]}")

embarked_counts = titanic_df["Embarked"].value_counts()
print(f"embarked_counts shape: {embarked_counts.shape}")

tot_embarked_count = 0
for embarked_count in embarked_counts:
    tot_emabrked_count += embarked_count

print(f"embarked 데이터 갯수: {tot_emabrked_count}")

타이타닉 데이터 갯수: 891
embarked_counts shape: (3,)
embarked 데이터 갯수: 889


전체 데이터 891개에서 2개가 부족하다는 것을 알 수 있습니다.<br>
하지만 위의 `info()`를 보면 891개라고 표시되어 있어 데이터 갯수가 일치하지 않습니다.

이 2개의 데이터는 NaN 값이며, 다른 언어의 null과 같습니다.<br>
기본적으로 `value_counts()`는 NaN 값을 제외하고 표시하기 때문에 모든 값을 표시하고 싶다면 인자에 `dropna=False`를 사용해서 표시할 수 있습니다.

In [33]:
embarked_with_nan_counts = titanic_df["Embarked"].value_counts(dropna=False)

print(embarked_counts)
print(embarked_with_nan_counts)

Embarked
S    644
C    168
Q     77
Name: count, dtype: int64
Embarked
S      644
C      168
Q       77
NaN      2
Name: count, dtype: int64


### DataFrame의 형변환

DataFrame은 `read_csv()`처럼 csv파일에서도 데이터를 읽어 초기화 할 수도 있지만<br>
Python의 list, dictionary, Numpy의 ndarray 등에서도 초기화 할 수 있습니다.<br>
물론 반대의 경우도 변환이 가능합니다.

API마다 입력인자로 받는 객체가 다르기 때문에 (특히 ndarray와 DataFrame 중 하나일 경우가 많습니다.)<br>
Pandas에서 이러한 형태의 형변환을 지원하게 됩니다.

#### ndarray, List, Dict -> DataFrame

위 collection과 DataFrame의 차이는 index가 명시적으로 선언되어 있는지의 차이가 있습니다.<br>
다만 DataFrame은 2차원 배열만을 저장할 수 있기 때문에, 다차원 ndarray는 변환할 수 없습니다.

1차원 배열을 DataFrame으로 변환하려면 column 지정이 필요하며 `columns=`를 인자로 넣어 지정합니다.

In [36]:
# Data type을 명시적으로 표시하기 위한 내장 패키지
from typing import List, Dict
import numpy as np

col_name1: List = ["col1"]
list1: List = [1, 2, 3]

# list를 ndarray로 형변환
arr1: np.ndarray = np.array(list1)

# 리스트로 DataFrame 생성
# list1은 1차원이기 때문에 columns로 열 지정
list1_df: pd.DataFrame = pd.DataFrame(list1, columns=col_name1)
print(f"list1_df shape: {list1_df.shape}")
print(f"list_df:\n{list1_df}")

# ndarray로 DataFrame 생성
print(f"arr1 shape: {arr1.shape}")
# arr1도 1차원이기 때문에 columns로 열 지정
arr1_df: pd.DataFrame = pd.DataFrame(arr1, columns=col_name1)
print(f"arr1_df:\n{arr1_df}")

list1_df shape: (3, 1)
list_df:
   col1
0     1
1     2
2     3
arr1 shape: (3,)
arr1_df:
   col1
0     1
1     2
2     3


2차원 배열또한 column을 명시적으로 선언해줍니다.

In [40]:
# 2x3 2차원 배열 생성
col_name2: List = ["col1", "col2", "col3"]
list2: List = [[1, 2, 3],
         [11, 12, 13]]

list2_df: pd.DataFrame = pd.DataFrame(list2, columns=col_name2)
print(f"list2_df:\n{list2_df}")

arr2: np.ndarray = np.array(list2)
print(f"arr2 shape: {arr2.shape}")

arr2_df: pd.DataFrame = pd.DataFrame(arr2, columns=col_name2)
print(f"arr2_df:\n{arr2_df}")


list2_df:
   col1  col2  col3
0     1     2     3
1    11    12    13
arr2 shape: (2, 3)
arr2_df:
   col1  col2  col3
0     1     2     3
1    11    12    13


Dictionary의 경우 Key를 컬럼명으로, Value는 데이터로 변환됩니다.<br>
ndarray는 Dictionary 변환을 지원하지 않습니다.

In [44]:
dict1: Dict = {"col1": [1, 2, 3, 4],
         "col2": [11, 12, 13, 14],
         "col3": [21, 22, 23, 24]}

dict1_df = pd.DataFrame(dict1)
print(f"dict1_df:\n{dict1_df}")

dict1_df:
   col1  col2  col3
0     1    11    21
1     2    12    22
2     3    13    23
3     4    14    24


#### DataFrame -> ndarray, List, Dict

반대의 경우도 당연히 가능합니다.<br>
특히 DataFrame을 ndarray로 변환하는 과정은 실제로 가장 많이 쓰이게 될 기능이라 익숙해지면 좋습니다.<br>
DataFrame의 `values`로 컬럼과 index를 제외한 값을 shape에 맞게 가져올 수 있습니다.

List의 경우 ndarray의 `tolist()` 함수로 변환하여 총 2번 바꾸게 됩니다.

Dictionary의 경우 DataFrame에서 `to_dict()` 함수를 사용하고, 인자로 'list'를 넣어 해당 값을 list형으로 가져옵니다.

In [47]:
# dict1_df는 4x3 형태의 2차원 배열
## ndarray 변환
arr3: np.ndarray = dict1_df.values
print(f"arr3 type: {type(arr3)}")
print(f"arr3 shape: {arr3.shape}")
print(f"arr3:\n{arr3}")

## List 변환
list3: List = dict1_df.values.tolist()
print(f"list3 type: {type(list3)}")
print(f"list3:\n{list3}")

## Dictionary 변환
dict3: Dict = dict1_df.to_dict("list")
print(f"dict3 type: {type(dict3)}")
print(f"dict3:\n{dict3}")

arr3 type: <class 'numpy.ndarray'>
arr3 shape: (4, 3)
arr3:
[[ 1 11 21]
 [ 2 12 22]
 [ 3 13 23]
 [ 4 14 24]]
list3 type: <class 'list'>
list3:
[[1, 11, 21], [2, 12, 22], [3, 13, 23], [4, 14, 24]]
dict3 type: <class 'dict'>
dict3:
{'col1': [1, 2, 3, 4], 'col2': [11, 12, 13, 14], 'col3': [21, 22, 23, 24]}


## DataFrame의 Column 추가 및 수정

DataFrame은 mutable한 객체이므로, 컬럼이나 데이터 수정이 자유롭습니다.

### Column 추가

`[]`로 Column을 추가할 수 있으며, 초기값을 지정해 줄 수 있습니다.

In [57]:
# 원본 유지를 위해 복사합니다.
## deep은 깊은 복사를 수행합니다.
titanic_cp_df = titanic_df.copy(deep=True)

# Age_0 컬럼을 초기값 0으로 생성
titanic_cp_df["Age_0"] = 0
# 기존 컬럼을 사용하여 생성
titanic_cp_df["Age_by_10"] = titanic_cp_df["Age"] * 10
titanic_cp_df["Family_No"] = titanic_cp_df["SibSp"] + titanic_cp_df["Parch"] + 1

titanic_cp_df.head(3)

,PassengerId,Survived,Pclass,Name,Sex,Age,SibSp,Parch,Ticket,Fare,Cabin,Embarked,Age_0,Age_by_10,Family_No
0,1,0,3,"Braund, Mr. Owen Harris",male,22.0,1,0,A/5 21171,7.2500,NaN,S,0,220.0,2
1,2,1,1,"Cumings, Mrs. John Bradley (Florence Briggs Th...",female,38.0,1,0,PC 17599,71.2833,C85,C,0,380.0,2
2,3,1,3,"Heikkinen, Miss. Laina",female,26.0,0,0,STON/O2. 3101282,7.9250,NaN,S,0,260.0,1


#### 데이터 일괄 수정

해당 컬럼에 접근하여 여러 연산으로 컬럼 값을 덮어씌울 수 있습니다.

In [58]:
titanic_cp_df["Age_by_10"] = titanic_cp_df["Age_by_10"] + 100

titanic_cp_df.head(3)

,PassengerId,Survived,Pclass,Name,Sex,Age,SibSp,Parch,Ticket,Fare,Cabin,Embarked,Age_0,Age_by_10,Family_No
0,1,0,3,"Braund, Mr. Owen Harris",male,22.0,1,0,A/5 21171,7.2500,NaN,S,0,320.0,2
1,2,1,1,"Cumings, Mrs. John Bradley (Florence Briggs Th...",female,38.0,1,0,PC 17599,71.2833,C85,C,0,480.0,2
2,3,1,3,"Heikkinen, Miss. Laina",female,26.0,0,0,STON/O2. 3101282,7.9250,NaN,S,0,360.0,1


### DataFrame 데이터 삭제

`drop()` 함수를 사용하여 행이나 열을 삭제합니다.<br>
주로 쓰이는 인자는 다음과 같습니다.
- `axis=` :<br>
대부분의 경우 행을 지우는 경우가 없기 때문에 (데이터 n건을 삭제하는 효과) 필요없는 column을 삭제하기 위해서 `axis=1`를 사용하게 됩니다.<br>
이상치에 해당하는 데이터를 삭제하는 경우 `axis=0`을 사용합니다.
- `inplace=` :<br>
데이터 삭제는 risk가 있는 작업일 수 밖에 없기에 기본적으로는 삭제된 DataFrame을 복사해서 반환합니다.<br>
`inplace=True`가 설정되면, 원본에서 직접 삭제하게 됩니다.
- `labels=` :<br>
여러 행, 열을 삭제하고 싶을 때 리스트로 index나 column name을 지정하여 사용합니다.

위에서 생성한 모든 컬럼을 삭제하도록 하겠습니다.

In [59]:
# column 삭제
need_drop_cols: List = ["Age_0", "Age_by_10", "Family_No"]
# inplace가 True 이므로, 반환값은 None이 됩니다.
drop_result: pd.DataFrame = titanic_cp_df.drop(need_drop_cols, axis=1, inplace=True)
print(f"drop_result:\n{drop_result}")
print("원본:")
titanic_cp_df.head(3)

drop_result:
None
원본:


,PassengerId,Survived,Pclass,Name,Sex,Age,SibSp,Parch,Ticket,Fare,Cabin,Embarked
0,1,0,3,"Braund, Mr. Owen Harris",male,22.0,1,0,A/5 21171,7.2500,NaN,S
1,2,1,1,"Cumings, Mrs. John Bradley (Florence Briggs Th...",female,38.0,1,0,PC 17599,71.2833,C85,C
2,3,1,3,"Heikkinen, Miss. Laina",female,26.0,0,0,STON/O2. 3101282,7.9250,NaN,S


In [60]:
# row 삭제
need_drop_rows = [0, 1, 2]
titanic_cp_df.drop(labels=need_drop_rows, axis=0, inplace=True)
titanic_cp_df.head(3)

,PassengerId,Survived,Pclass,Name,Sex,Age,SibSp,Parch,Ticket,Fare,Cabin,Embarked
3,4,1,1,"Futrelle, Mrs. Jacques Heath (Lily May Peel)",female,35.0,1,0,113803,53.1000,C123,S
4,5,0,3,"Allen, Mr. William Henry",male,35.0,0,0,373450,8.0500,NaN,S
5,6,0,3,"Moran, Mr. James",male,NaN,0,0,330877,8.4583,NaN,Q


## Index 객체

Index는 DataFrame에서 PK의 역할을 하는 객체입니다.<br>
`index` 로 따로 배열로 추출할 수 있습니다.<br>
`values`로 ndarray 형태로 추출할 수 있어 numpy의 API를 모두 사용할 수 있습니다.

index 자체는 Immutable 객체이기 때문에 직접 변경이 불가능합니다.

In [63]:
# cp 원복
titanic_cp_df: pd.DataFrame = titanic_df.copy(deep=True)
# index 배열 추출
indexes: pd.RangeIndex = titanic_cp_df.index
print(f"indexes type: {type(indexes)}")
print(f"indexes shape: {indexes.shape}")
print(f"indexes:\n{indexes}")

# array로 변환
index_arr: np.ndarray = indexes.values
print(f"index_arr type: {type(index_arr)}")
print(f"index_arr shape: {index_arr.shape}")
print(f"index arr:\n{indexes.values}")


indexes type: <class 'pandas.core.indexes.range.RangeIndex'>
indexes shape: (891,)
indexes:
RangeIndex(start=0, stop=891, step=1)
index_arr type: <class 'numpy.ndarray'>
index_arr shape: (891,)
index arr:
[  0   1   2   3   4   5   6   7   8   9  10  11  12  13  14  15  16  17
  18  19  20  21  22  23  24  25  26  27  28  29  30  31  32  33  34  35
  36  37  38  39  40  41  42  43  44  45  46  47  48  49  50  51  52  53
  54  55  56  57  58  59  60  61  62  63  64  65  66  67  68  69  70  71
  72  73  74  75  76  77  78  79  80  81  82  83  84  85  86  87  88  89
  90  91  92  93  94  95  96  97  98  99 100 101 102 103 104 105 106 107
 108 109 110 111 112 113 114 115 116 117 118 119 120 121 122 123 124 125
 126 127 128 129 130 131 132 133 134 135 136 137 138 139 140 141 142 143
 144 145 146 147 148 149 150 151 152 153 154 155 156 157 158 159 160 161
 162 163 164 165 166 167 168 169 170 171 172 173 174 175 176 177 178 179
 180 181 182 183 184 185 186 187 188 189 190 191 192 193 194 195 

`RangeIndex` 객체는 ndarray처럼 slicing이 가능합니다.

In [65]:
# RangedIndex slicing
print(indexes[:5].values)
# ndarray slicing
print(indexes.values[:5])

[0 1 2 3 4]
[0 1 2 3 4]


Series 객체는 Index 객체를 포함하지만, 연산에는 사용되지 않습니다.<br>
따라서 Index는 식별용(참조용)으로만 사용하게 됩니다.

만약 index값이 연속된 숫자형태가 아니라면, 데이터 sort기준으로 새로 인덱스를 부여해줄 수 있습니다.<br>
`reset_index()`로 새로 생성할 수 있고 `inplace=True`가 기본값입니다.<br>
`inplace=False`인 경우, 기존 index를 컬럼으로 추가하고 새로운 연속 숫자형 index를 생성합니다.

In [66]:
titanic_cp_df = titanic_cp_df.reset_index(inplace=False)
titanic_cp_df.head(3)

,index,PassengerId,Survived,Pclass,Name,Sex,Age,SibSp,Parch,Ticket,Fare,Cabin,Embarked
0,0,1,0,3,"Braund, Mr. Owen Harris",male,22.0,1,0,A/5 21171,7.2500,NaN,S
1,1,2,1,1,"Cumings, Mrs. John Bradley (Florence Briggs Th...",female,38.0,1,0,PC 17599,71.2833,C85,C
2,2,3,1,3,"Heikkinen, Miss. Laina",female,26.0,0,0,STON/O2. 3101282,7.9250,NaN,S


이전 Pclass Series를 가져왔을 때 [3, 1, 2]로 구분되어 있었는데 여기서 `reset_index()`를 실행하면 기존 [3, 1, 2]가 column으로 바뀌고,<br>
연속 [0, 1, 2] 값이 index로 지정되게 됩니다.<br>
`inplace=False` 라면 column이 추가되었으므로 Data type은 DataFrame으로 자동으로 바뀝니다.

In [67]:
pclass_series = titanic_cp_df["Pclass"].value_counts()
print(f"pclass_series type: {type(pclass_series)}")
print(f"pclass_series:\n{pclass_series}")

# index 새로 지정
pclass_df = pclass_series.reset_index(inplace=False)
print(f"pclass_df type: {type(pclass_df)}")
print(f"pclass_df:\n{pclass_df}")

pclass_series type: <class 'pandas.core.series.Series'>
pclass_series:
Pclass
3    491
1    216
2    184
Name: count, dtype: int64
pclass_df type: <class 'pandas.core.frame.DataFrame'>
pclass_df:
   Pclass  count
0       3    491
1       1    216
2       2    184


## 데이터 셀렉션 및 필터링

Pandas는 Numpy와 상당히 유사한 셀렉션 및 필터링 기능을 가지고 있지만, 다른 부분도 분명히 있으므로 혼동되기 쉽습니다.<br>
Numpy의 데이터 핸들링은 데이터 분석용으로는 한계가 있기 때문에, Pandas에서는 Numpy의 기능을 계승하거나 혹은 완전히 다른 기능을 제공하기도 합니다.<br>
DataFrame과 Series끼리도 달라지는 부분이 있어서 주의가 필요합니다.

Numpy에서는 `[]` 연산자를 사용하여 다음과 같은 기능을 수행했습니다.
- 단일 값 추출
- Slicing
- Fancy Indexing
- Boolean Indexing

하지만 Pandas에서는 이러한 작업을 `loc[]`, `iloc[]` 연산자를 통해 수행하게 됩니다.<br>

먼저 Numpy의 `[]`와 Pandas의 `[]`의 차이를 확인해보도록 하겠습니다.

### DataFrame의 `[]`연산자

Numpy에서 `[]`연산자는 행의 위치, 열의 위치, Slicing 범위 등을 지정해서 가져올 수 있었습니다.<br>
하지만 DataFrame의 `[]`연산자는 column명, 또는 index로 변환 가능한 표현식(statement)입니다.<br>

In [68]:
# df 원복
titanic_cp_df = titanic_df.copy(deep=True)

print(f"단일 column 데이터 추출:\n{titanic_cp_df['Pclass'].head(3)}")
print(f"여러 column 데이터 추출:\n{titanic_cp_df[['Survived', 'Pclass']].head(3)}")

# index로 접근시 에러 발생
print(f"index 로 접근:\n{titanic_cp_df[0]}")

단일 column 데이터 추출:
0    3
1    1
2    3
Name: Pclass, dtype: int64
여러 column 데이터 추출:
   Survived  Pclass
0         0       3
1         1       1
2         1       3


KeyError: 0

인덱스 형태로 변환 가능한 표현식은 다음 요소만 가능합니다.
- Slicing
- Boolean Indexing

거의 대부분이 Boolean Indexing을 활용하므로, 익숙해지면 좋습니다.

In [69]:
# Slicing
# head(3)와 같습니다.
titanic_cp_df[0:3]

,PassengerId,Survived,Pclass,Name,Sex,Age,SibSp,Parch,Ticket,Fare,Cabin,Embarked
0,1,0,3,"Braund, Mr. Owen Harris",male,22.0,1,0,A/5 21171,7.2500,NaN,S
1,2,1,1,"Cumings, Mrs. John Bradley (Florence Briggs Th...",female,38.0,1,0,PC 17599,71.2833,C85,C
2,3,1,3,"Heikkinen, Miss. Laina",female,26.0,0,0,STON/O2. 3101282,7.9250,NaN,S


In [70]:
# Boolean Indexing
# Pclass가 3인 최상위 3개
titanic_cp_df[titanic_cp_df["Pclass"] == 3].head(3)

,PassengerId,Survived,Pclass,Name,Sex,Age,SibSp,Parch,Ticket,Fare,Cabin,Embarked
0,1,0,3,"Braund, Mr. Owen Harris",male,22.0,1,0,A/5 21171,7.250,NaN,S
2,3,1,3,"Heikkinen, Miss. Laina",female,26.0,0,0,STON/O2. 3101282,7.925,NaN,S
4,5,0,3,"Allen, Mr. William Henry",male,35.0,0,0,373450,8.050,NaN,S


### `iloc[]` 연산자

`iloc[]`의 경우 인덱스 기반으로 DataFrame의 값을 참조하는 함수입니다.<br>
따라서 숫자가 아닌 값으로 참조하면 에러가 발생합니다.

In [71]:
# 임시 DataFrame 생성
data = {"name": ["A", "B", "C", "D"],
        "age": [35, 23, 50, 15],
        "gender": ["M", "F", "F", "M"]}
data_df = pd.DataFrame(data, index=["one", "two", "three", "four"])
data_df

,name,age,gender
one,A,35,M
two,B,23,F
three,C,50,F
four,D,15,M


In [72]:
# 인덱스 기반으로 참조
data_df.iloc[0, 0]

'A'

In [73]:
# 숫자가 아닌 값을 지정하면 에러
data_df.iloc[0, "B"]

ValueError: Location based indexing can only have [integer, integer slice (START point is INCLUDED, END point is EXCLUDED), listlike of integers, boolean array] types

`iloc[]`은 숫자형을 사용하기 때문에 다음과 같은 연산 또한 가능합니다.

In [74]:
# 단일 값 반환
data_df.iloc[0, 1]

35

In [77]:
# 슬라이싱 범위의 행에서 해당 열 리스트에 속하는 열을 반환
data_df.iloc[0:2, [0, 2]]

,name,gender
one,A,M
two,B,F


In [79]:
# 슬라이싱 범위의 행에서 슬라이싱 범위의 열 반환
data_df.iloc[0:3, 1:3]

,age,gender
one,35,M
two,23,F
three,50,F


In [80]:
# 전체 반환
data_df.iloc[:]

,name,age,gender
one,A,35,M
two,B,23,F
three,C,50,F
four,D,15,M


In [81]:
# 마지막 column 반환
data_df.iloc[:, -1]

one      M
two      F
three    F
four     M
Name: gender, dtype: object

In [82]:
# 마지막 column을 제외한 모든 데이터
data_df.iloc[:, :-1]

,name,age
one,A,35
two,B,23
three,C,50
four,D,15


### `loc[]`연산자

`loc[]`은 Label 기반으로 데이터를 추출합니다.<br>
형식은 [(index), (column명)] 입니다.<br>
여기서는 index를 `str`로 지정하였으므로 문자열로 indexing 합니다.

In [84]:
# 1행의 Name은?
data_df.loc["one", "name"]

'A'

숫자 기반 Indexing에서 Slicing은 마지막 값을 포함하지 않지만, Label 기반에서는 마지막 Label을 포함합니다.<br>
또한 Label 기반이기에 `-1`을 사용할 수 없습니다.

In [86]:
print(f"위치 기반 iloc slicing:\n{data_df.iloc[0:1, 0]}")
# 1번 index를 포함하여 출력됩니다.
print(f"라벨 기반 loc slicing:\n{data_df.loc['one':'two', 'name']}")

위치 기반 iloc slicing:
one    A
Name: name, dtype: object
라벨 기반 loc slicing:
one    A
two    B
Name: name, dtype: object


In [87]:
# 단일 값 반환
data_df.loc["one", "name"]

'A'

In [88]:
# 행의 범위에 속하는 특정 열 반환
data_df.loc["one":"three", ["name", "age"]]

,name,age
one,A,35
two,B,23
three,C,50


In [90]:
# 행의 범위에 속하는 열의 범위 반환
data_df.loc["one":"two", "name":"gender"]

,name,age,gender
one,A,35,M
two,B,23,F


In [91]:
# 모든 데이터 값
data_df.loc[:]

,name,age,gender
one,A,35,M
two,B,23,F
three,C,50,F
four,D,15,M


In [94]:
# Boolean Indexing
# iloc과 다르게 가능합니다.
data_df.loc[data_df.age > 30, ["name", "gender"]]

,name,gender
one,A,M
three,C,F


### Boolean Indexing

Pandas는 `iloc[]`을 제외하고는 `[]`, `loc[]` 모두 Boolean Indexing을 지원합니다.

In [96]:
titanic_age_mask = titanic_cp_df["Age"] > 60
over_sixty_age = titanic_cp_df[titanic_age_mask]

print(f"mask type: {type(titanic_age_mask)}")
print(f"mask shape: {titanic_age_mask.shape}")
print(f"result type: {type(over_sixty_age)}")
print(f"result shape: {over_sixty_age.shape}")
over_sixty_age

mask type: <class 'pandas.core.series.Series'>
mask shape: (891,)
result type: <class 'pandas.core.frame.DataFrame'>
result shape: (22, 12)


,PassengerId,Survived,Pclass,Name,Sex,Age,SibSp,Parch,Ticket,Fare,Cabin,Embarked
33,34,0,2,"Wheadon, Mr. Edward H",male,66.0,0,0,C.A. 24579,10.5000,NaN,S
54,55,0,1,"Ostby, Mr. Engelhart Cornelius",male,65.0,0,1,113509,61.9792,B30,C
96,97,0,1,"Goldschmidt, Mr. George B",male,71.0,0,0,PC 17754,34.6542,A5,C
116,117,0,3,"Connors, Mr. Patrick",male,70.5,0,0,370369,7.7500,NaN,Q
170,171,0,1,"Van der hoef, Mr. Wyckoff",male,61.0,0,0,111240,33.5000,B19,S
252,253,0,1,"Stead, Mr. William Thomas",male,62.0,0,0,113514,26.5500,C87,S
275,276,1,1,"Andrews, Miss. Kornelia Theodosia",female,63.0,1,0,13502,77.9583,D7,S
280,281,0,3,"Duane, Mr. Frank",male,65.0,0,0,336439,7.7500,NaN,Q
326,327,0,3,"Nysveen, Mr. Johan Hansen",male,61.0,0,0,345364,6.2375,NaN,S
438,439,0,1,"Fortune, Mr. Mark",male,64.0,1,4,19950,263.0000,C23 C25 C27,S


또한 복합적인 조건문도 결합해서 사용할 수 있습니다.
- &: and
- |: or
- ~: not

In [98]:
# 조건 결합
## 1. 나이가 60세 이상
## 2. 등급이 1등급
## 3. 성별이 여성

age_mask = titanic_cp_df["Age"] >= 60
pclass_mask = titanic_cp_df["Pclass"] == 1
sex_mask = titanic_cp_df["Sex"] == "female"

result = titanic_cp_df[age_mask & pclass_mask & sex_mask]
result

,PassengerId,Survived,Pclass,Name,Sex,Age,SibSp,Parch,Ticket,Fare,Cabin,Embarked
275,276,1,1,"Andrews, Miss. Kornelia Theodosia",female,63.0,1,0,13502,77.9583,D7,S
366,367,1,1,"Warren, Mrs. Frank Manley (Anna Sophia Atkinson)",female,60.0,1,0,110813,75.2500,D37,C
829,830,1,1,"Stone, Mrs. George Nelson (Martha Evelyn)",female,62.0,0,0,113572,80.0000,B28,NaN


## 정렬, Aggregation, GroupBy

### DataFrame, Series의 정렬 - sort_values()

두 객체 모두 `sort_values()`를 지원합니다.<br>
이 함수는 RDBMS의 `order by` 키워드와 매우 유사합니다.<br>
주요 파라미터는 다음과 같습니다.
- by: 특정 column을 기준으로 정렬을 수행합니다.
- ascending: 오름차순, 내림차순 여부를 결정합니다. [True/False] 기본값은 True입니다.
- inplace: ndarray와 같이 원본 적용 여부를 결정합니다. [True/False] 기본값은 False 입니다.

In [99]:
# 단일 column으로 정렬
titanic_sorted_asc = titanic_cp_df.sort_values(by="Name")
titanic_sorted_asc.head(3)

,PassengerId,Survived,Pclass,Name,Sex,Age,SibSp,Parch,Ticket,Fare,Cabin,Embarked
845,846,0,3,"Abbing, Mr. Anthony",male,42.0,0,0,C.A. 5547,7.55,NaN,S
746,747,0,3,"Abbott, Mr. Rossmore Edward",male,16.0,1,1,C.A. 2673,20.25,NaN,S
279,280,1,3,"Abbott, Mrs. Stanton (Rosa Hunt)",female,35.0,1,1,C.A. 2673,20.25,NaN,S


In [101]:
# 복수의 column으로 정렬
## 먼저 지정된 column 기준으로 정렬됩니다.
titanic_sorted_asc = titanic_cp_df.sort_values(by=["Pclass", "Name"], ascending=False)
titanic_sorted_asc.head(3)

,PassengerId,Survived,Pclass,Name,Sex,Age,SibSp,Parch,Ticket,Fare,Cabin,Embarked
868,869,0,3,"van Melkebeke, Mr. Philemon",male,NaN,0,0,345777,9.5,NaN,S
153,154,0,3,"van Billiard, Mr. Austin Blyler",male,40.5,0,2,A/5. 851,14.5,NaN,S
282,283,0,3,"de Pelsmaeker, Mr. Alfons",male,16.0,0,0,345778,9.5,NaN,S


### Aggregation

RDBMS의 aggregation 함수와 유사하게 작동합니다.<br>
주요 aggregation 함수는 다음과 같습니다.
- min(): 해당 열의 최솟값을 출력합니다.
- max(): 해당 열의 최댓값을 출력합니다.
- sum(): 해당 열의 합산을 출력합니다.
- count(): 해당 열의 갯수를 출력합니다.
    - count의 경우 결측치를 제외하고 계산합니다.
- mean(): 해당 열의 평균값을 출력합니다.

In [102]:
titanic_cp_df.count()

PassengerId    891
Survived       891
Pclass         891
Name           891
Sex            891
Age            714
SibSp          891
Parch          891
Ticket         891
Fare           891
Cabin          204
Embarked       889
dtype: int64

특정 column에 aggregation을 적용하고 싶다면 column을 추출해서 계산하면 됩니다.

In [103]:
titanic_cp_df[["Age", "Fare"]].mean()

Age     29.699118
Fare    32.204208
dtype: float64

### groupby()

RDBMS의 `group by`와 비슷하지만 다른 면이 있기 때문에 주의해야 합니다.<br>
DataFrame에서 `groupby()`를 호출하게 되면 DataFrameGroupBy라는 다른 형태의 DataFrame을 반환하게 됩니다.

In [104]:
titanic_groupby = titanic_cp_df.groupby(by="Pclass")
print(type(titanic_groupby))

<class 'pandas.core.groupby.generic.DataFrameGroupBy'>


DataFrame에서 `groupby()`를 수행하면 지정한 column을 제외하고 aggregation이 적용됩니다.<br>
RDBMS에서는 모든 column에 aggregation을 적용해야 하지만 `SELECT count(PassengerId), count(Survived),...`<br>
DataFrame은 일괄 적용 시킨다는 점이 다르다고 할 수 있습니다.

In [105]:
# Aggregation 일괄 적용
titanic_groupby = titanic_cp_df.groupby(by="Pclass").count()
titanic_groupby

,PassengerId,Survived,Name,Sex,Age,SibSp,Parch,Ticket,Fare,Cabin,Embarked
Pclass,,,,,,,,,,,
1,216,216,216,216,186,216,216,216,216,176,214
2,184,184,184,184,173,184,184,184,184,16,184
3,491,491,491,491,355,491,491,491,491,12,491


In [108]:
# 특정 column에만 Aggregation 적용
titanic_groupby = titanic_cp_df.groupby(by="Pclass")[["PassengerId", "Survived"]].count()
titanic_groupby

,PassengerId,Survived
Pclass,,
1,216,216
2,184,184
3,491,491


만약 하나의 컬럼에 여러 aggregation을 적용하고 싶거나 여러 컬럼에 각각의 aggregation을 적용하고 싶다면 `agg()` 함수를 사용합니다.<br>
`agg()` 함수는 aggregation list와 "column": "aggregation" 형태의 dictionary를 받을 수 있습니다.<br>
확실히 RDBMS보다는 복잡한 과정을 가지고 있습니다.

In [110]:
# aggregation list
## Pclass 별 Age의 최소, 최대 출력
print("Pclass별 Age 최대 최소")
titanic_cp_df.groupby("Pclass")["Age"].agg([max, min])

Pclass별 Age 최대 최소


,max,min
Pclass,,
1,80.0,0.92
2,70.0,0.67
3,74.0,0.42


In [112]:
# aggregation dictionary
agg_format = {"Age": "max", "SibSp": "sum", "Fare": "mean"}
print("Pclass 별 Age 최대, SibSp 총합, Fare 평균")
titanic_cp_df.groupby("Pclass").agg(agg_format)

Pclass 별 Age 최대, SibSp 총합, Fare 평균


,Age,SibSp,Fare
Pclass,,,
1,80.0,90,84.154687
2,70.0,74,20.662183
3,74.0,302,13.675550


## 결측치 처리

결측치는 cell에 데이터가 없는 경우를 의미하며, Numpy의 NaN으로 표시합니다.<br>
ML은 이 결측치를 자동으로 처리하지 않으므로 다른 값으로 채워넣어야 합니다.<br>
또한 이 결측치는 Aggregation 시 제외가 되므로 데이터 신뢰도가 낮아지는 현상이 있을 수 있습니다.

### 결측치 여부 확인 - isna()

isna()는 해당 데이터가 NaN인지를 알려주는 함수입니다.<br>
DataFrame에 적용하면 모든 column의 값이 NaN인지 아닌지를 [True/False]로 반환합니다.

In [113]:
titanic_cp_df.isna().head()

,PassengerId,Survived,Pclass,Name,Sex,Age,SibSp,Parch,Ticket,Fare,Cabin,Embarked
0,False,False,False,False,False,False,False,False,False,False,True,False
1,False,False,False,False,False,False,False,False,False,False,False,False
2,False,False,False,False,False,False,False,False,False,False,True,False
3,False,False,False,False,False,False,False,False,False,False,False,False
4,False,False,False,False,False,False,False,False,False,False,True,False


In [114]:
# 결측치 갯수 확인
titanic_cp_df.isna().sum()

PassengerId      0
Survived         0
Pclass           0
Name             0
Sex              0
Age            177
SibSp            0
Parch            0
Ticket           0
Fare             0
Cabin          687
Embarked         2
dtype: int64

### 결측치 대체 - fillna()

결측치는 일일히 넣어줄 필요없이 `fillna()` 함수로 일괄 처리할 수 있습니다.<br>
이 함수 역시 `inplace` 인자가 존재하며, 기본값은 False 입니다.

In [115]:
# Cabin column의 결측치를 모두 C000으로 대체합니다.
titanic_cp_df["Cabin"] = titanic_cp_df["Cabin"].fillna("C000")
titanic_cp_df.head(3)

,PassengerId,Survived,Pclass,Name,Sex,Age,SibSp,Parch,Ticket,Fare,Cabin,Embarked
0,1,0,3,"Braund, Mr. Owen Harris",male,22.0,1,0,A/5 21171,7.2500,C000,S
1,2,1,1,"Cumings, Mrs. John Bradley (Florence Briggs Th...",female,38.0,1,0,PC 17599,71.2833,C85,C
2,3,1,3,"Heikkinen, Miss. Laina",female,26.0,0,0,STON/O2. 3101282,7.9250,C000,S


In [117]:
# 1. Age의 결측치를 모두 평균값으로 대체
# 2. Embarked의 결측치를 모두 S로 대체
age_mean = titanic_cp_df["Age"].mean()
print(f"Age의 평균값: {age_mean}")

titanic_cp_df["Age"].fillna(age_mean, inplace=True)
titanic_cp_df["Embarked"].fillna("S", inplace=True)

# 모든 결측치가 처리되었습니다.
titanic_cp_df.isna().sum()

Age의 평균값: 29.69911764705882


PassengerId    0
Survived       0
Pclass         0
Name           0
Sex            0
Age            0
SibSp          0
Parch          0
Ticket         0
Fare           0
Cabin          0
Embarked       0
dtype: int64

## lambda 식으로 데이터 가공

lambda 식은 대부분의 경우 익숙하지 않은 표현식이지만, 언어와 상관없이 사용되는 개념입니다.<br>
Python에서의 lambda 문법이긴 하지만 DataFrame에서는 `apply()`로 lambda식을 활용할 수 있습니다.

lambda에 대해 모르겠으면 Python을 다시 공부하고 오시면 됩니다. ^^

In [118]:
# 예시 함수 선언
def get_square(a):
    return a ** 2

print(f"3의 제곱은 {get_square(3)}")

3의 제곱은 9


In [119]:
lambda_square = lambda x: x ** 2
print(f"3의 제곱은 {lambda_square(3)}")

3의 제곱은 9


Python에서는 여러 값에 일괄적용하기 위해서는 `map()`을 사용해야 합니다.

In [121]:
list1 = [1, 2, 3]
list1_squares = list(map(lambda_square, list1))
print(list1_squares)

[1, 4, 9]


Pandas에서는 이러한 lambda 식을 적용하기 위해 `apply()` 함수를 사용합니다.

In [122]:
titanic_cp_df["Name_len"] = titanic_cp_df["Name"].apply(lambda x: len(x))
titanic_cp_df[["Name", "Name_len"]].head(3)

,Name,Name_len
0,"Braund, Mr. Owen Harris",23
1,"Cumings, Mrs. John Bradley (Florence Briggs Th...",51
2,"Heikkinen, Miss. Laina",22


Python의 lambda식 형식에 맞다면, 복잡한 구문도 적용이 가능합니다.<br>
다만 너무 길어지면 가독성이 매우 떨어지므로, 별도의 함수로 구현하는게 좋습니다.

In [126]:
# 18세 이하라면 Child, 그 이외는 Adult
titanic_cp_df["Child_Adult"] = titanic_cp_df["Age"].apply(lambda x: "Child" if x <= 18 else "Adult")
titanic_cp_df[["Age", "Child_Adult"]].head(8)

,Age,Child_Adult
0,22.000000,Adult
1,38.000000,Adult
2,26.000000,Adult
3,35.000000,Adult
4,35.000000,Adult
5,29.699118,Adult
6,54.000000,Adult
7,2.000000,Child


In [127]:
titanic_cp_df["Child_Adult"].value_counts()

Child_Adult
Adult    752
Child    139
Name: count, dtype: int64

In [128]:
# 함수로 lambda 적용
def get_category(age):
    cat = ""
    if age <= 5: cat = "Baby"
    elif age <= 12: cat = "Child"
    elif age <= 18: cat = "Teenager"
    elif age <= 25: cat = "Student"
    elif age <= 35: cat = "Young Adult"
    elif age <= 60: cat = "Adult"
    else: cat = "Elderly"

    return cat

titanic_cp_df["Age_cat"] = titanic_cp_df["Age"].apply(lambda x: get_category(x))
titanic_cp_df["Age_cat"].value_counts()

Age_cat
Young Adult    373
Adult          195
Student        162
Teenager        70
Baby            44
Child           25
Elderly         22
Name: count, dtype: int64